crates/environments/src/loader.rs

```diff
@@ -0,0 +1,211 @@
+use std::path::Path;
+
+use anyhow::{anyhow, Context};
+use spin_common::ui::quoted_path;
+
+pub(crate) struct ComponentToValidate<'a> {
+    id: &'a str,
+    source_description: String,
+    wasm: Vec<u8>,
+}
+
+struct ComponentSource<'a> {
+    id: &'a str,
+    source: &'a spin_manifest::schema::v2::ComponentSource,
+    dependencies: WrappedComponentDependencies,
+}
+
+impl<'a> ComponentToValidate<'a> {
+    pub fn id(&self) -> &str {
+        self.id
+    }
+
+    pub fn source_description(&self) -> &str {
+        &self.source_description
+    }
+
+    pub fn wasm_bytes(&self) -> &[u8] {
+        &self.wasm
+    }
+}
+
+pub async fn load_and_resolve_all<'a>(
+    app: &'a spin_manifest::schema::v2::AppManifest,
```

Comment:
Nit: seems like we pass `app` and `resolution_context` around a lot. It might be nicer to put those into a `Resolver` struct and implement these functions as methods on that struct.

Reply:
I had an explore of this, and have pushed something like it as a separate commit for easy reversion.  I ended up naming the proposed struct `ApplicationToValidate` because otherwise I ended up with no "application" visible in a module that was ostensibly all about validating applications.  I still have mixed feelings but there's definitely some nice aspects to encapsulating the component loading stuff - I was able to tuck some more clutter away and split things up in a way that's hopefully more readable...